source/differentiation/ad/ad_helpers.cc

```diff
@@ -0,0 +1,783 @@
+
+      stream << "Touched independent variables: "
+             << "\n";
+      for (unsigned int i = 0; i < n_independent_variables(); i++)
+        stream << touched_independent_variables[i]
+               << (i < (n_independent_variables() - 1) ? "," : "");
+      stream << std::endl;
+
+      stream << "Independent variable values: "
+             << "\n";
+      print_values(stream);
+
+      stream << "Touched marked independent variables: "
+             << "\n";
+      for (unsigned int i = 0; i < n_independent_variables(); i++)
+        stream << touched_marked_independent_variables[i]
+               << (i < (n_independent_variables() - 1) ? "," : "")
+               << std::flush;
+      stream << std::endl;
+
+      stream << "Dependent variable values: "
+             << "\n";
+      for (unsigned int i = 0; i < n_dependent_variables(); i++)
+        stream << dependent_variables[i]
+               << (i < (n_dependent_variables() - 1) ? "," : "") << std::flush;
+      stream << std::endl;
+
+      stream << "Touched dependent variables: "
+             << "\n";
+      for (unsigned int i = 0; i < n_dependent_variables(); i++)
+        stream << touched_dependent_variables[i]
+               << (i < (n_dependent_variables() - 1) ? "," : "") << std::flush;
+      stream << std::endl;
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    void
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::print_values(
+      std::ostream &stream) const
+    {
+      for (unsigned int i = 0; i < n_independent_variables(); i++)
+        stream << independent_variable_values[i]
+               << (i < (n_independent_variables() - 1) ? "," : "")
+               << std::flush;
+      stream << std::endl;
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    void
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::print_tape_stats(
+      std::ostream &           stream,
+      const types::tape_index &tape_index) const
+    {
+      if (ADNumberTraits<ad_type>::is_tapeless == true)
+        return;
+
+      Assert(is_registered_tape(tape_index),
+             ExcMessage("Tape number not registered"));
+
+      TapedDrivers<ad_type, scalar_type>::print_tape_stats(stream, tape_index);
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    void
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::reset(
+      const unsigned int n_independent_variables,
+      const unsigned int n_dependent_variables,
+      const bool         clear_registered_tapes)
+    {
+      const unsigned int new_n_independent_variables =
+        (n_independent_variables != 0 ? n_independent_variables :
+                                        this->n_independent_variables());
```

Comment:
I didn't pay attention when I read the function docs, but are you using zero as a special value here? I think that's not a great choice. Why not use `numbers::invalid_unsigned_int` as everywhere else?